ESP scanning test

filipleple opened the pull request.

Commit:
dasharo-security/option-rom.rom: added test

Signed-off-by: Mixss <michal.ziemiec@3mdeb.com>

Commit:
lib/bios/menus.robot: added Read options from list kw

Signed-off-by: Mixss <michal.ziemiec@3mdeb.com>

filipleple committed:
dasharo-compatibility/esp-scanning.robot: added

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

Psotas committed:
rebase: move keyword to lib/bios/menus.robot

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
lib/bios/menus.robot add Enter Dasharo Submenu keyword

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
add: boot-maintenance-manager.robot

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
fix: codespell & robotidy

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

filipleple commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+    [Documentation]    Mounts a tiny bootable ISO as a flash stick for testing purposes.
+
+    Upload Image To PiKVM    ${PIKVM_IP}
+    ...    https://cloud.3mdeb.com/index.php/s/Gckspfa5A6TT9fK/download/CorePlus.iso
```
@miczyg1 We could, but I thought we have more certainty that a link to our cloud gives us more control/certainty that it doesn't change

macpijan replied:
@filipleple It makes some sense, but let's use official URL at first and see how well it works. Please use the full URL so it is also clear which ISO version we use.

filipleple replied:
Fixed in [recent commit](https://github.com/Dasharo/open-source-firmware-validation/commit/429cccd6960bb98ab4f3dfb8dc6ba01028ce30e1)

filipleple commented on platform-configs/msi-pro-z690-a-wifi-ddr4.robot:
```diff
@@ -55,10 +51,10 @@ ${DEVICE_WINDOWS_USERNAME}=                         user
+${DEVICE_UBUNTU_USERNAME}=                          ubuntu
+${DEVICE_UBUNTU_PASSWORD}=                          ubuntu
+${DEVICE_UBUNTU_USER_PROMPT}=                       ubuntu@3mdeb:
+${DEVICE_UBUNTU_ROOT_PROMPT}=                       root@3mdeb:
```
Fixed in [recent commit](https://github.com/Dasharo/open-source-firmware-validation/commit/429cccd6960bb98ab4f3dfb8dc6ba01028ce30e1)

Psotas committed:
add variables for all platform

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

miczyg1 commented on platform-configs/msi-pro-z690-a-wifi-ddr4.robot:
```diff
@@ -55,10 +55,10 @@ ${DEVICE_WINDOWS_USERNAME}=                         user
-${DEVICE_UBUNTU_PASSWORD}=                          ${UBUNTU_PASSWORD}
-${DEVICE_UBUNTU_USER_PROMPT}=                       ${UBUNTU_USER_PROMPT}
-${DEVICE_UBUNTU_ROOT_PROMPT}=                       ${UBUNTU_ROOT_PROMPT}
+${DEVICE_UBUNTU_USERNAME}=                          ---
+${DEVICE_UBUNTU_PASSWORD}=                          ---
+${DEVICE_UBUNTU_USER_PROMPT}=                       ---
+${DEVICE_UBUNTU_ROOT_PROMPT}=                       ---
```
Why do we change it? Can't it be left as it was before using the UBUNTU_ variables?

filipleple replied:
@miczyg1 sorry, change removed

miczyg1 commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+
+    Upload Image To PiKVM    ${PIKVM_IP}
+    ...    https://distro.ibiblio.org/tinycorelinux/14.x/x86/release/CorePlus-current.iso
+    Mount Image On PiKVM    ${PIKVM_IP}    CorePlus.iso
```
It will be CorePlus-current.iso AFAIK

filipleple replied:
Right, fixed [here](https://github.com/Dasharo/open-source-firmware-validation/pull/31/files#diff-1c553d8d90529e6578e4c3b7bec1aea478c26edce6d99d353b8a35b02c05c67cR126)

miczyg1 commented on dasharo-compatibility/esp-scanning.robot:
```diff
@@ -0,0 +1,98 @@
+    Upload And Mount DTS Flash ISO
+    Power On
+    Enter Boot Menu Tianocore
+    Check That USB Devices Are Detected    ${TRUE}
```
Where is the check for `Dasharo Tools Suite (on <disk_name>)`?

filipleple replied:
@miczyg1 Right - IIRC on removable media DTS doesn't get listed as `Dasharo Tools Suite (on <disk_name>)` like it does on the SSD, but gets listed as the flash drive brand name or `PiKVM Composite Device`. I should rather report this as a bug than make the test go with it.

filipleple replied:
@miczyg1 I rewrote the check for DTS [here](https://github.com/Dasharo/open-source-firmware-validation/pull/31/files#diff-c8f7a5cdc7c64ba736049ef48cf813ae918ab342659418e21d720981c53f903aR99). It always fails due to what I mentioned above. AFAIK the esp-scan was a feature merged to our edk2 fork - I wanted to report this issue, but there is no issues tab under edk2. Where should I report the issue then?

miczyg1 replied:
I think @TomaszAIR didn't fix it yet. The DTS boot partition is standard FAT32, not EFI System Partition type. That is why it cannot be found.

macpijan commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+    Should Be Equal As Integers    ${x}    ${n}
+
+Remove All Supported Systems From Efi
+    Remove System From Efi    Fedora
```
How about a list and iterate through that with a loop instead?

filipleple replied:
Rewritten [here](https://github.com/Dasharo/open-source-firmware-validation/pull/31/files#diff-1c553d8d90529e6578e4c3b7bec1aea478c26edce6d99d353b8a35b02c05c67cR103)

macpijan commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+
+Remove System From Efi
+    [Arguments]    ${folder_name}
+    SSHLibrary.Execute Command    rm -r /boot/efi/EFI/${folder_name}
```
This is executed on the DUT?
We should not use SSHLibrary directly for this. See: https://github.com/Dasharo/open-source-firmware-validation#guidelines

filipleple replied:
All instances replaced with Execute Command In Terminal

macpijan commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+Add System Folder To /EFI
+    [Documentation]    This keyword creates directories and .efi files for
+    ...    given systems. Available options:
+    ...    ${type} - linux, windows, manager
```
What `manager` stands for?

filipleple replied:
It's to handle the boot managers - they use elilo.efi instead of grubx64/shimx64. Will clarify that more in the keyword's documentation

macpijan commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+    ...    empty - keyword fails if the list is not empty
+    [Arguments]    ${mode}=normal
+
+    ${boot_list}=    Get Boot Menu Construction
```
This logic is quite convoluted for me TBH. Do you have the full expected Boot Menu Construction with the supported OSes?

filipleple replied:
There's a little bit more to it. AFAIK a system not being listed vs being listed more than once are fairly different issues so I've split them into 2 test cases. In the normal mode we verify that the OSes show up at all, in the double_entry_check mode we verify that there's only one entry per OS - I thought having 2 keywords which mostly overlap would be redundant, so I came up with one keyword with settable modes. Though I see it's fairly complicated and I can split them if we favor readability.

macpijan committed:
.github: move starting qemu into qemu-run.sh script

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
README.md: add info on QEMU

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

filipleple commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+    ...    populating ESP with all the required files.
+    ...    Available modes:
+    ...    normal
+    ...    empty - keyword fails if the list is not empty
```
In one case we check the boot menu for all the OSes, in the other we expect the boot menu to be empty - I thought it'd be more efficient to have that in one keyword rather than two mostly overlapping keywords

Commit:
dasharo-security/option-rom.robot: updated test's documentation

Signed-off-by: Mixss <michal.ziemiec@3mdeb.com>

filipleple commented on lib/esp-scanning-lib.robot:
```diff
@@ -0,0 +1,126 @@
+    END
+    SSHLibrary.Execute Command    wget ${GRUBX64_URL} -O ${folder_name}/grubx64.efi
+
+    # since switch to root user doesn't really wanna work
```
Right, now working on reporting the issue

macpijan committed:
README.md: link to getting started in QEMU section

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

Psotas committed:
remove:

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

macpijan committed:
working 1st kwds

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
2nd milestone with passing self-tests

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
list options + working networking-boot on MSI

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
scripts/ci/qemu-run.sh: update links to QEMU

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
boot-maintenance-manager.robot: working test in qemu

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
self-tests: update regexp

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

Psotas committed:
dasharo-security/me-neuter.robot: change to new kwds

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
dasharo-security/smm-bios-write-protection.robot: change to new kwds

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
dasharo-security/bios-lock.robot: change to new kwds

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
dasharo-security/early-boot-dma-protection.robot: change to new kwds

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
dasharo-security/usb-stack.robot: change to new kwds

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

macpijan committed:
add refactoring hints and script

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

Psotas committed:
remove kwds: Read Option List Contents

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Check If Tianocore Setting Is Enabled In Current Menu

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Change To Next Option In Setting

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Check If Submenu Exists Tianocore

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Change Numeric Value Of Setting

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Enter Dasharo Submenu Snapshot

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Enter Device Manager Submenu

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Enter Secure Boot Configuration Submenu

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Tianocore One Time Boot

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

Psotas committed:
remove kwds: Enter UEFI Shell Tianocore

Signed-off-by: Piotr Lasota <piotr.lasota@3mdeb.com>

pietrushnic committed:
README.md: fix QEMU as supported platform

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
improve qemy-run.sh script and its documentation

Make sure qemy-run.sh script can be used for validation and development
purposes by adding parameter that handle custom paths of OVMF_CODE and
OVMF_VARS. Extend script to inform users what is happening. Extend
documentation to describe behavior and improve copy pastablility.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: add support for vnc

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu.robot: enable network boot testing

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu.robot: set IPXE_BOOT_ENTRY

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
dasharo-compatibility/network-boot.robot: make sure network is enabled before testing

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
platform-configs/qemu.robot: set correct value for EDK2_IPXE_CHECKPOINT

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
dasharo-compatibility/network-boot.robot: fix PXE002.001

Because of changes in Get IPXE Boot Menu Construction, which cut off
first line of menu we have to expect differnt lines.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
scripts/ci/qemu-run.sh: apply minimum amount of RAM to boo DTS

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
dasharo-compatibility/network-boot.robot: remove netwokr enabling

Remove network enabling from PXE001.001. If one would like to achieve
correct result one should run test that enables network stack before
test that test iPXE e.g. NBA001.001.

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
dasharo-compatibility/network-boot.robot: fix DTS menu in  PXE004.001

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
.github/workflows/qemu-self-test.yml: fix qemu-run.sh params

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

pietrushnic committed:
fix some pre-commit issues

Signed-off-by: Piotr Król <piotr.krol@3mdeb.com>

macpijan committed:
keys-and-keywords/firmware-keywords.robot: remove unused file

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
keys-and-keywords/setup-keywords.robot: remove unused file

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
remove kwds: Enter Submenu In Tianocore

1. netwok-boot.robot tested on QEMU
2. netwok-boot-utilities.robot tested on VP4630

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
remove kwds: unmark Reenter Menu

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
remove kwds: Get Relative Menu Position

It was no longer used anywhere

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
remove kwds: Get Menu Reference Tianocore

It was no longer used anywhere

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
custom-boot-menu-key.robot: fix Enter Boot Menu kwd

Enter Boot Menu does not exist anymore

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
handle parsing of option values spanning into two lines

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
dasharo-compatibility/network-boot.robot: MSI fixes

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
platform-configs: msi: update variables

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
dasharo-compatibility/memory-profile.robot: adjust to reworked keywords

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

macpijan committed:
dasharo-compatibility/reset-to-defaults.robot: udpate with reworked kwds

Signed-off-by: Maciej Pijanowski <maciej.pijanowski@3mdeb.com>

filipleple committed:
qemu edition wip

Signed-off-by: Filip Lewiński <filip.lewinski@3mdeb.com>

filipleple commented:
Replaced by #88
